`crux-mir-test`: Improve output for failing tests

At the moment, if a test case fails in `crux-mir`'s test suite, it looks something like this:

```
      checked_add:               FAIL (1.26s)
        files test/symb_eval/num/checked_add.good and test/symb_eval/num/checked_add.out differ; test/symb_eval/num/checked_add.out contains:
        test checked_add/<DISAMB>::crux_test[0]: returned 44, FAILED

        failures:

        ---- checked_add/<DISAMB>::crux_test[0] counterexamples ----
        [Crux] Found counterexample for verification goal
        [Crux]   test/symb_eval/num/checked_add.rs:6:5: 6:12: error: in checked_add/<DISAMB>::crux_test[0]
        [Crux]   Overflow(Add, const 100_u8, move _1)

        [Crux] Overall status: Invalid.

        Use -p '/checked_add/' to rerun this test only.
```

This output is rather poor, as it doesn't indicate precisely _why_ the output differs. Is the output completely different? Only different in one line? It's impossible to say without manually opening the golden file (`checked_add.good`) and comparing it to the new output, which is tedious.

Compare this to the [output of a failed test](https://github.com/GaloisInc/crucible/pull/1173#issuecomment-1944787931) in `crux-llvm`'s test suite:

```
                T972-fail crux output:                   FAIL
                  test/Test.hs:310:
                  MISMATCH for expected (test-data/golden/T972-fail.pre-clang14.z3.good)
                             and actual (test-data/golden/T972-fail.pre-clang14.z3.z3.out) output:
                      F        |[Crux] Found counterexample for verification goal
                      F        |[Crux]   test-data/golden/T972-fail.c:8:3: error: in main
                      F-expect>|[Crux]   unsupported LLVM value: ValAsm True False "testl $0, $0; jne ${1:l};" "r,X,X,~{dirflag},~{fpsr},~{flags}" of type void(i32, i8*, i8*)*
                      F-actual>|[Crux]   unsupported LLVM value: ValAsm True False "testl $0, $0; jne ${1:l};" "r,X,X" of type void(i32, i8*, i8*)*
                      F-expect>|[Crux] Overall status: Invalid.
                      F-actual>|[Crux] Failed to build counterexample executable
                      <snip>
```

This shows both the golden test's output and the new output side by side, and it reveals that the output only differs starting on the third line. We should do something similar in `crux-mir`'s test suite.

--- a/crux-mir/test/Test.hs
+++ b/crux-mir/test/Test.hs
@@ -6,6 +6,7 @@
 module Main (main) where
 
 import qualified Data.ByteString as BS
+import qualified Data.ByteString.Char8 as BSC
 import qualified Data.ByteString.UTF8 as BS8
 import           Data.Char (isSpace)
 import           Data.List (dropWhileEnd, isPrefixOf)
@@ -169,8 +170,27 @@ doGoldenTest rustFile goodFile outFile act = goldenTest (takeBaseName rustFile)
     (BS.readFile goodFile)
     (act >> BS.readFile outFile)
     (\good out -> return $ if good == out then Nothing else
-      Just $ "files " ++ goodFile ++ " and " ++ outFile ++ " differ; " ++
-        outFile ++ " contains:\n" ++ BS8.toString out)
+      let el = BSC.lines good
+          al = BSC.lines out in
+
+      let dl (e,a) = if e == a then db e else de e <> da a
+          db b = ["    F        |" <> b]
+          de e = ["    F-expect>|" <> e]
+          da a = ["    F-actual>|" <> a] in
+
+      let details = concat $
+                    [ [ "MISMATCH for expected (" <> BS8.fromString goodFile <> ")"
+                      , "           and actual (" <> BS8.fromString outFile <> ") output:"
+                      ]
+                      -- Highly simplistic "diff" output assumes
+                      -- correlated lines: added or removed lines just
+                      -- cause everything to shown as different from
+                      -- that point forward.
+                    , concatMap dl $ zip el al
+                    , concatMap de $ drop (length al) el
+                    , concatMap da $ drop (length el) al
+                    ] in
+      Just $ BS8.toString (BSC.unlines details))
     (\out -> BS.writeFile goodFile out)
 
 -- | Sanitize the golden output in a file.

